control-plane/connect-inject/controllers/endpoints/endpoints_controller.go

```diff
@@ -1336,6 +1345,28 @@ func isGateway(pod corev1.Pod) bool {
 	return ok && anno != ""
 }
 
+// isTelemetryCollector checks whether a pod is part of a deployment for a Consul Telemetry Collector. If so,
+// and this is the first pod deployed to a Namespace, we need to create the Namespace in Consul. Otherwise the
+// deployment may fail out during service registration because it is deployed to a Namespace that does not exist.
+func isTelemetryCollector(pod corev1.Pod) bool {
+	anno, ok := pod.Annotations[constants.LabelTelemetryCollector]
+	return ok && anno != ""
+}
+
+// ensureNamespaceExists creates a Consul namespace for a pod in the event it does not exist.
+// At the time of writing, we use this for the Consul Telemetry Collector which may be the first
+// pod deployed to a namespace. If it is, it's connect-inject will fail for lack of a namespace.
+func (r *Controller) ensureNamespaceExists(apiClient *api.Client, pod corev1.Pod) error {
+	if r.EnableConsulNamespaces {
+		consulNS := r.consulNamespace(pod.Namespace)
+		if _, err := namespaces.EnsureExists(apiClient, consulNS, r.CrossNSACLPolicy); err != nil {
```

Comment:
this is copy-pasta from the registerGateways method above: https://github.com/hashicorp/consul-k8s/blob/e7e31d4a8a814bf8e6ceb2a26ff3827b969d1362/control-plane/connect-inject/controllers/endpoints/endpoints_controller.go#L355